### DISCUSSION TITLE
Clash in Vulkan "CreateInstance()" Times

### DISCUSSION THREAD

**Discussion by kgs233** [dotnet/Silk.NET#1844] [Q&A] [answered]
### What happened？
When I was trying to create a Vulkan instance, I ran into a strange issue where it would crash without error from time to time.
Specifically:
```
Vk vk = Vk.GetApi();
...
vk.CreateInstance(createInfo, null, out Instance pInstance); //There
...
```
After that, the program just exits without leaving any error message (but this is only in the terminal and in vscode, in vs it will report an error "heap memory corruption")
In Debug mode the program returns 0xC0000374, which is shown in the windows log tracker as "STATUS_HEAP_CORRUPTION".

### What might have happened?
I' dont knowI think the problem is with P/Invoke though, maybe the GC is reclaiming resources that shouldn't be reclaimed? But I don't know how to fix it.

### Where's the weird part?
This problem doesn't always happen, sometimes it works fine in debug mode but not in release mode, sometimes it works fine in the debugger but not outside of it, and even sometimes it still occurs from time to time when you run the same version of the program in the same environment.

**Comment by Beyley** [comment]
this error suggests you are writing out of bounds somewhere in your code, or you are passing bad pointers into the driver, and it is writing out of bounds, we'd need more code reference to know whether or not this is our problem or yours.

**Reply by Perksey** [reply]
Yep. We'll likely need more of a reproduction than this to escalate to a bug.

**Reply by kgs233** [reply]
```
using Silk.NET.Core.Native;
using Silk.NET.SDL;
using Silk.NET.Vulkan;
using System.Runtime.InteropServices;
using static SDL2.SDL;

namespace A;
public class A
{
    public static unsafe void Main(string[] args)
    {
        Vk vk = Vk.GetApi();

        IntPtr enginWindow = SDL_CreateWindow("Test",
                                     SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED,
                                     800, 900,
                                     SDL_WindowFlags.SDL_WINDOW_SHOWN |
                                     SDL_WindowFlags.SDL_WINDOW_VULKAN);

        ApplicationInfo appInfo = new()
        {
            SType = StructureType.ApplicationInfo,
            PApplicationName = (byte*) Marshal.StringToHGlobalAnsi("abc"),
            ApplicationVersion = 0,
            EngineVersion = 0,
            PEngineName = (byte*) Marshal.StringToHGlobalAnsi("Gal Engin"),
            ApiVersion = Vk.Version11
        };

        SDL_Vulkan_GetInstanceExtensions(enginWindow , out uint SDLExtensionCount, null);
        IntPtr SDLExtensions = SilkMarshal.AllocateString((int)SDLExtensionCount);
        SDL_Vulkan_GetInstanceExtensions(enginWindow, out SDLExtensionCount, SDLExtensions);
        string[] extensions = SilkMarshal.PtrToStringArray(SDLExtensions, (int)SDLExtensionCount);

        InstanceCreateInfo createInfo = new()
        {
            SType = StructureType.InstanceCreateInfo,
            PApplicationInfo = &appInfo,
            EnabledExtensionCount = (uint)extensions.Length,
            PpEnabledExtensionNames = (byte**)SilkMarshal.StringArrayToPtr(extensions),
            EnabledLayerCount = 0,
            PNext = null
        };
        createInfo.EnabledLayerCount = 0;
        createInfo.PNext = null;

        vk.CreateInstance(createInfo, null, out Instance pInstance);

        Marshal.FreeHGlobal((nint) appInfo.PApplicationName);
        Marshal.FreeHGlobal((nint) appInfo.PEngineName);
        SilkMarshal.Free((nint) createInfo.PpEnabledLayerNames);
    }
}
```
**Even with this code, an exception occurs**

**Reply by Perksey** [reply]
Sorry, can't reproduce here on macOS - I'm getting a valid instance handle.

**Reply by Perksey** [reply]
Okay I can reproduce on Windows, investigating. This is likely not our fault so not escalating yet.

**Reply by Perksey** [reply]
I can't get anywhere with this, the issue won't reproduce when I've got WinDbg attached.

**Reply by kgs233** [reply]
Indeed, since I also tried running it in WinDbg, but I couldn't reproduce the problem

**Reply by Perksey** [reply, marked as answer]
The problem appears to be
```cs
        IntPtr SDLExtensions = SilkMarshal.AllocateString((int)SDLExtensionCount);
```
This should be
```cs
        IntPtr SDLExtensions = SilkMarshal.Allocate((int)SDLExtensionCount * sizeof(nint));
```
You are attempting to write `SDLExtensionCount` pointers into memory of size `SDLExtensionCount`. So for example, if SDL wants to enable 2 extensions, you are attempting to write 8-16 bytes into a 2 byte buffer.

**Reply by Perksey** [reply]
^ special thanks to @tannergooding for helping figure this out

**Reply by kgs233** [reply]
It worked. Thank you.
